crates/bevy_ecs/src/world/archetype_invariants.rs

```diff
@@ -0,0 +1,730 @@
+    /// Although this appears asymmetric, it actually implies its own converse.
+    /// This is particularly useful for avoiding query conflicts.
+    #[inline]
+    fn forbids<B2: Bundle>() -> ArchetypeInvariant<B, B2> {
+        ArchetypeInvariant {
+            premise: ArchetypeStatement::<B>::any_of(),
+            consequence: ArchetypeStatement::<B2>::none_of(),
+        }
+    }
+
+    /// Creates an archetype invariant where components of `B` require all the components of `B2`.
+    ///
+    /// In other words, if any component from `B` is present, then all of the components from `B2` must be.
+    #[inline]
+    fn requires<B2: Bundle>() -> ArchetypeInvariant<B, B2> {
+        ArchetypeInvariant {
+            premise: ArchetypeStatement::<B>::any_of(),
+            consequence: ArchetypeStatement::<B2>::all_of(),
+        }
+    }
+
+    /// Creates an archetype invariant where components of `B` require at least one component of `B2`.
+    ///
+    /// In other words, if any component from `B` is present, then at least one component from `B2` must be.
+    #[inline]
+    fn requires_one<B2: Bundle>() -> ArchetypeInvariant<B, B2> {
+        ArchetypeInvariant {
+            premise: ArchetypeStatement::<B>::any_of(),
+            consequence: ArchetypeStatement::<B2>::any_of(),
+        }
+    }
+
+    /// Creates an archetype invariant where all components of `B` require each other.
+    ///
+    /// In other words, if any component of this bundle is present, then all of them must be.
+    #[inline]
+    fn atomic() -> ArchetypeInvariant<B> {
+        ArchetypeInvariant {
+            premise: ArchetypeStatement::<B>::any_of(),
+            consequence: ArchetypeStatement::<B>::all_of(),
+        }
+    }
+
+    /// Creates an archetype where components of `B` cannot appear with each other.
+    ///
+    /// In other words, if any component of this bundle is present, then no others can be.
+    /// This is particularly useful for creating enum-like groups of components, such as `Dead` and `Alive`.
+    #[inline]
+    fn disjoint() -> ArchetypeInvariant<B> {
+        ArchetypeInvariant {
+            premise: ArchetypeStatement::<B>::any_of(),
+            consequence: ArchetypeStatement::<B>::at_most_one_of(),
+        }
+    }
+
+    /// Creates an archetype invariant where components of `B` can only appear with each other.
+    ///
+    /// In other words, if any component of this bundle is present, then _only_ components from this bundle can be present.
+    #[inline]
+    fn exclusive() -> ArchetypeInvariant<B> {
+        ArchetypeInvariant {
+            premise: ArchetypeStatement::<B>::any_of(),
+            consequence: ArchetypeStatement::<B>::only(),
+        }
+    }
+}
+
+/// A special module used to prevent [`ArchetypeInvariantHelpers`] from being implemented by any other module.
+///
+/// For more information, see: <https://rust-lang.github.io/api-guidelines/future-proofing.html>.
```

Comment:
It might be worth it to link to the [Restrictions RFC](https://rust-lang.github.io/rfcs/3323-restrictions.html) here, which could (as far as I understand it) replace this pattern once implemented.